vllm/model_executor/layers/sampler.py

```diff
@@ -83,10 +113,93 @@ def forward(
         # Compute the probabilities.
         probs = torch.softmax(logits, dim=-1, dtype=torch.float)
         # Compute the log probabilities (before applying top-p and top-k).
-        logprobs = torch.log(probs)
+        # Re-compute to ensure numerical stability.
+        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
 
         # Sample the next tokens.
-        return _sample(probs, logprobs, input_metadata)
+        samples = _sample(probs, logprobs, input_metadata)
+        if echo_results:
+            for i, ss in enumerate(samples):
```

Comment:
Please avoid confusing variable names like `ss`. Please rename it to `sample` for example.